Ensure user-provided ConfigMaps are not modified

mgoerens opened the pull request:
This commit adds the creation of configMaps for the homeserver.yaml and
heisenbridge.yaml configuration files, even if the user has provided
such a file. We now consider that user-provided ConfigMaps are not to be
owned (and therefore modified) by the controller.

closes #17

acmenezes commented:
I'll check that out as soon as I get another time slot. There are too many changes in a single commit.

acmenezes commented on controllers/synapse/synapse_configmap.go:
```diff
@@ -2677,9 +2677,33 @@ redis:
+	return cm, nil
+}
+
+func (r *SynapseReconciler) configMapForSynapseCopy(
+	s *synapsev1alpha1.Synapse,
+	objectMeta metav1.ObjectMeta,
+) (client.Object, error) {
+	var copyConfigMap *corev1.ConfigMap
+
+	sourceConfigMapName := s.Spec.Homeserver.ConfigMap.Name
+	sourceConfigMapNamespace := r.getConfigMapNamespace(*s, s.Spec.Homeserver.ConfigMap.Namespace)
+
+	copyConfigMap, err := r.getConfigMapCopy(sourceConfigMapName, sourceConfigMapNamespace, objectMeta)
+	if err != nil {
+		return &corev1.ConfigMap{}, err
+	}
```
On the attempt of understanding the code here I see two functions pretty similar in name `configMapForSynapseCopy` and `getConfigMapCopy`. Some comments here can help clarify the differences between the two and turn the code more explicit. Changing the names may also make sense.

mgoerens replied:
I added some comments, let me know if that makes it clearer

acmenezes commented on controllers/synapse/synapse_controller.go:
```diff
@@ -70,8 +70,10 @@ type HomeserverPgsqlDatabase struct {
 
-	// Load the Synapse by name
 	var synapse synapsev1alpha1.Synapse
+	var createdConfigMap corev1.ConfigMap
```
Here is another name that made me a bit confused. I have the feeling that without proper comments and explanation I won't understand much what is the goal here. We have a pre-existent ConfigMap, a Copied ConfigMap and also a createdConfigMap. That may be ok if those three versions are needed but I'm not reading it straightforward on the code. I got a bit lost here.

acmenezes replied:
Awesome! That's way better.

acmenezes commented on controllers/synapse/synapse_controller.go:
```diff
@@ -94,35 +96,61 @@ func (r *SynapseReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
+			ctx,
+			r.configMapForSynapseCopy,
+			&synapse,
+			&createdConfigMap,
```
Since the code is a bit long for this file, wouldn't be a good idea bringing the createdConfigMap variable declaration closer to where it's being used?

mgoerens replied:
I followed your suggestion and moved the variable declaration closer. Note that this is used in both branch of the if/else block as follow:


```go
	var createdConfigMap corev1.ConfigMap

	if synapse.Spec.Homeserver.ConfigMap != nil {
		// createdConfigMap becomes a copy of the existing ConfigMap
	} else {
		// A new ConfigMap is created and stored in createdConfigMap
	}
```

acmenezes replied:
It's clear now.

acmenezes reviewed (commented):
Hi @mgoerens , I'm trying to get my feet wet again on this project. Just saw your last commit here and wanted to scan over the code and give you feedback. The code looks sane but I got confused with some names. Put my comments on that. And there is way more changes to this than the title of this PR reveals. As a general recommendation I would break it in multiple well explained commits next times. But happy to help! We hop into a call some time during the week to catch up on this if needed.

mgoerens commented:
> Hi @mgoerens , I'm trying to get my feet wet again on this project. Just saw your last commit here and wanted to scan over the code and give you feedback. The code looks sane but I got confused with some names. Put my comments on that. And there is way more changes to this than the title of this PR reveals. As a general recommendation I would break it in multiple well explained commits next times. But happy to help! We hop into a call some time during the week to catch up on this if needed.

Hey @acmenezes ! First off, thanks for taking a look. Some context on the changes added in this MR:

The main goal, as described in the commit message, is to ensure that the ConfigMaps created by the user (not by the controller) remain the user's ownership. Until now, they were modified by the controller (to add some information such as a url, or database connection information). Now we create a copy of the user-provided ConfigMap where we can work freely.

### Context

We have to work with multiple ConfigMaps, and all their related functions are located in `synapse_*configmap.go` files:
- The Synapse configuration file (`homeserver.yaml`), defined in `synapse_configmap.go`
- and the Heisenbridge config file (`heisenbridge.yaml`), defined in `synapse_heisenbridge_configmap.go`
- (In another branch, I'm also working on the Signal Bridge, which comes with a `config.yaml` config file.)

In addition, we have the `configmap_utils.go` for generic functions related to ConfigMap manipulation.

### Previously

In a previous commit (0ba537e49b705af4d8e22423240e4dd82ecfa236), we added generic functions for loading, updating, and writing to a ConfigMap. FYI (this is not part of this MR), I went with a similar pattern as for the `reconcile.go` code (which comes from you :) ). The `updateConfigMap` function takes an argument of type `updateDataFunc`. In each `synapse_*configmap.go` files I added corresponding functions. For now we have:
- `updateHomeserverWithHeisenbridgeInfos` in `synapse_configmap.go`.
- `updateHeisenbridgeWithURL` in `synapse_heisenbridge_configmap.go`
- (and soon we'll have something similar for Signal)

### In this MR

Now, in this MR (finally), we add the "copy" feature. For creating a ConfigMap from an existing one (copy), I leveraged the existing `reconcileResource` function, which takes a `createResourceFunc` function as an argument. Similarly, I created corresponding functions in `synapse_*configmap.go`:
- `configMapForHeisenbridgeCopy` in `synapse_heisenbridge_configmap.go`
- and `configMapForSynapseCopy` in `synapse_configmap.go`

(That also explains why there are some changes to `reconcile.go` which seem unrelated to the rest. I had to slightly modify the signature of the `createResourceFunc` to fit the needs of the `*Copy` functions. Not very clear, sorry)

Both those functions rely on a generic `getConfigMapCopy` function, located in `configmap_utils.go` for all the common bits of copying an existing ConfigMap.

------------------

Hope that helps a little ! Thanks for your comments. I agree that I could have done a better job at splitting my work into a few commits. There are some changes which are a bit difficult to understand without context (the changes to `reconcile.go` for instance) I try to get better, I promise :)

I will address your comments and attempt to clarify the code with comments / better naming.

mgoerens commented:
urg, sorry I realized I already squashed the fixup commit adding comments to the changes in this PR.

In addition, there is a separate commit for comments to other parts of the code.

acmenezes commented on controllers/synapse/configmap_utils.go:
```diff
@@ -130,3 +131,29 @@ func (r *SynapseReconciler) writeYAMLFileToConfigMapData(
+		Data:       sourceConfigMap.Data,
+	}
+
+	return copyConfigMap, nil
```
Thanks!
